### ISSUE TITLE
KerasCV Fix-it ⚒️

### ISSUE THREAD

**Issue by divyashreepathihalli** [keras-team/keras-cv#2211]
Implement backend agnostic augmentation layers

- [ ] AugMix
- [ ] AutoContrast
- [ ] ChannelShuffle
- [ ] CutMix
- [ ] Equalization
- [ ] FourierMix
- [ ] Grayscale
- [ ] GridMask
- [ ] JitteredResize
- [ ] MixUp
- [ ] Mosaic
- [ ] Posterization
- [ ] RandAugment
- [ ] RandomApply
- [ ] RandomAspectRatio
- [ ] RandomBrightness
- [ ] RandomChannelShift
- [ ] RandomChoice
- [ ] RandomColorDegeneration
- [ ] RandomColorJitter
- [ ] RandomContrast
- [ ] RandomCropAndResize
- [ ] RandomCrop
- [ ] RandomCutout
- [ ] RandomFlip
- [ ] RandomGaussianBlur
- [ ] RandomHue
- [ ] RandomJpegQuality
- [ ] RandomRotation
- [ ] RandomSaturation
- [ ] RandomSharpness
- [ ] RandomShear
- [ ] RandomTranslation
- [ ] RandomZoom
- [ ] RepeatedAugmentation
- [ ] Rescaling
- [ ] Resizing
- [ ] Solarization

Implement the following layers in Keras 3

- [ ] keras_cv/layers/preprocessing_3d/base_augmentation_layer_3d.py
- [ ] frustum_random_dropping_points
- [ ] frustum_random_point_feature_noise.py
- [ ] global_random_dropping_points.py
- [ ] global_random_flip.py
- [ ] global_random_rotation.py
- [ ] global_random_scaling.py
- [ ] global_random_translation.py
- [ ] group_points_by_bounding_boxes.py
- [ ] random_copy_paste.py
- [ ] random_drop_box.py
- [ ] swap_background.py

 Update Keras IO examples

- [ ] learnable_resizer
- [ ] deit
- [ ] mixup
- [ ] perceiver_image_classification
- [ ] shiftvit
- [ ] bit
- [ ] simsiam
- [ ] generative/dcgan_overriding_train_ste
- [ ] forwardforward
- [ ] nnclr
- [ ] handwriting_recognition
- [ ] xray_classification_with_tpus
- [ ] edsr
- [ ] mobilevit
- [ ] yolov8
- [ ] video_classification
- [ ] nl_image_search
- [ ] supervised-contrastive-learning
- [ ] depth_estimation
- [ ] fully_convolutional_network
- [ ] vivit
- [ ] randaugment
- [ ] consistency_training
- [ ] basnet_segmentation
- [ ] barlow_twins
- [ ] adamatch
- [ ] pointnet
- [ ] vit_small_ds
- [ ] retinanet
- [ ] temporal_latent_bottleneck
- [ ] near_dup_search
- [ ] 3D_image_classification
- [ ] swin_transformers
- [ ] focal_modulation_network

**Comment by ID6109** [comment]
I'm excited to take these on. I'll be sending PRs on these after a week. Thanks!

**Comment by soma2000-lang** [comment]
Will work on a few of these @ID6109 which ones will you be working on?

**Comment by divyashreepathihalli** [comment]
@soma2000-lang and @ID6109 Thank you for you interest. I wanted to add an update. We have decided to keep the preprocessing layers TF only for now. So, I have updated the list above.

**Comment by ID6109** [comment]
I see @divyashreepathihalli. Then perhaps it was good that we waited a bit to initiate the conversions. @soma2000-lang, how about you start from the bottom? I'll begin by starting from the top.

**Comment by soma2000-lang** [comment]
Sure @ID6109 works for me.

**Comment by sami-bg** [comment]
`perceiver_image_classification` uses the LAMB optimizer, implemented in [TensorFlow Addons](https://www.tensorflow.org/addons/api_docs/python/tfa/optimizers/LAMB), as recommended from the paper. tfa is deprecated, so this particular example would need a keras-3 reimplementation of LAMB.

@soma2000-lang @ID6109 I'll be starting from the middle

**Comment by divyashreepathihalli** [comment]
@sami-bg yeah, let us leave the examples with custom implementation to the last.

**Comment by sachinprasadhs** [comment]
@divyashreepathihalli , For Handwriting recognition, `ctc_batch_cost` and `ctc_decode` is deprecated, any alternative that can be used?

**Comment by sachinprasadhs** [comment]
@sayakpaul , For `Near-duplicate image search` to migrate to `Keras 3` is it possible to provide the pre-trained model saved in `.keras` format here `bit_model = tf.keras.models.load_model("flower_model_bit_0.96875")`

**Comment by sachinprasadhs** [comment]
`Mixup` seems to be migrated to Keras 3 already  here https://github.com/keras-team/keras-io/pull/1595

**Comment by divyashreepathihalli** [comment]
Thanks Sachin! marked it as done

**Comment by sachinprasadhs** [comment]
I will work on `randaugment`

**Comment by divyashreepathihalli** [comment]
> I will work on `randaugment`

Thank you I will mark it as yours.

**Comment by divyashreepathihalli** [comment]
@IMvision12 in this example here - can you please add the details on what dataset you are using for the YOLO example here - https://keras.io/examples/vision/yolov8/
How to set it up and this example colab should run end to end ? - right now it cannot because of missing path to data.

**Comment by sineeli** [comment]
Hi @divyashreepathihalli,

I am trying to migrate [retinanet](https://keras.io/examples/vision/retinanet/) notebook, are there any alternatives that can be used for `tf.image` functions for example `tf.image.flip_right_left`. I tried but could find any related in keras_cv. 

Thanks

**Comment by divyashreepathihalli** [comment]
> Hi @divyashreepathihalli,
> 
> I am trying to migrate [retinanet](https://keras.io/examples/vision/retinanet/) notebook, are there any alternatives that can be used for `tf.image` functions for example `tf.image.flip_right_left`. I tried but could find any related in keras_cv.
> 
> Thanks

@sineeli you can use the preprocessing layers for that
for example
```
random_flip = keras_cv.layers.RandomFlip(rate=1.0, mode="horizontal")
fliped_image=random_flip(image)
```

We also have additional image ops here - https://keras.io/api/ops/image/

**Comment by sineeli** [comment]
@divyashreepathihalli it seems `keras_cv.layers. multi_class_non_max_suppression` does not support multi-backend  as per the code [here](https://github.com/keras-team/keras-cv/blob/ce88b46eac880417096343bb8ac6b0171f0baf85/keras_cv/layers/object_detection/multi_class_non_max_suppression.py#L76C2-L76C2) and retinanet notebook uses `tf.image.combined_non_max_suppression`.  it's parallel is `keras_cv.layers. multi_class_non_max_suppression`.

**Comment by IMvision12** [comment]
> @IMvision12 in this example here - can you please add the details on what dataset you are using for the YOLO example here - https://keras.io/examples/vision/yolov8/ How to set it up and this example colab should run end to end ? - right now it cannot because of missing path to data.

Hi @divyashreepathihalli you can use this to load data

```python
!gdown 19CWdNL3ePq9XIgMeb_RjA641iGra3nWm
!unzip data.zip
```

and update the code for loading the data

```python
# Path to images and annotations
path_images = "./data/images/"
path_annot = "./data/annotations/"
```

**Comment by divyashreepathihalli** [comment]
thank you @IMvision12!

**Comment by aditya02shah** [comment]
Hi @divyashreepathihalli 
I am migrating the [simsiam](https://github.com/keras-team/keras-io/blob/master/examples/vision/simsiam.py) example. It uses an implementation of [ResNet20](https://github.com/GoogleCloudPlatform/keras-idiomatic-programmer/blob/master/zoo/resnet/resnet_cifar10_v2.py) that will need to be updated too. 
Thanks

**Comment by sachinprasadhs** [comment]
Learnable resizer: https://github.com/keras-team/keras-io/pull/1688

**Comment by sitamgithub-MSIT** [comment]
I had worked on the bit example. One problem I faced was that the example is using the hub.keraslayer that is not compatible with keras 3. Anyway I tried using load the model .pb file with keras.layers.TFSMLayer it worked did not throw any error. But later in the notebook giving errors.

Any suggestions to solve it?

cc: @sachinprasadhs @divyashreepathihalli

**Comment by divyashreepathihalli** [comment]
@sitamgithub-MSIT it is okay to use `tensorflow_hub`. when you submit the PR to keras IO, please highlight that the example works only with TF backend.

**Comment by sitamgithub-MSIT** [comment]
@divyashreepathihalli I would like to take up the 3D_image_classification example.

**Comment by sitamgithub-MSIT** [comment]
> @divyashreepathihalli I would like to take up the 3D_image_classification example.

Getting this [ValueError: Cannot iterate over a shape with unknown rank] in model.fit(). Is there any way to resolve it? Or should I raise an issue about it!?

**Comment by sitamgithub-MSIT** [comment]
> > @divyashreepathihalli I would like to take up the 3D_image_classification example.
> 
> Getting this [ValueError: Cannot iterate over a shape with unknown rank] in model.fit(). Is there any way to resolve it? Or should I raise an issue about it!?

Issue solved. I will make a PR soon!

**Comment by divyashreepathihalli** [comment]
@sitamgithub-MSIT sorry about the late response. I am glad the issue was resolved.

**Comment by simeetnayan81** [comment]
Hey @sitamgithub-MSIT. Have you completed "3D_image_classification example"? If not I would like to take this up.

**Comment by simeetnayan81** [comment]
@divyashreepathihalli I will work on handwriting_recognition if that has not already been taken by someone else.

**Comment by sitamgithub-MSIT** [comment]
> Hey @sitamgithub-MSIT. Have you completed "3D_image_classification example"? If not I would like to take this up.

Yeahs it's done. Just I was busy so couldn't make a PR for that.

**Comment by sineeli** [comment]
Hey @divyashreepathihalli, If number of lines of the code exceeds 350 for existing example, what could be done? 

Please let me know if it's possible to migrate, thanks!

**Comment by divyashreepathihalli** [comment]
@sineeli you can temporarily update this https://github.com/keras-team/keras-io/blob/1d8d323165c345a176035af8a8044fa72595c28d/scripts/tutobooks.py#L73
to the desired limit and then change it back once you generate all the files.

**Comment by aditya02shah** [comment]
@divyashreepathihalli `SwinTransformers` seems to be migrated to Keras-3 already: https://github.com/keras-team/keras-io/pull/1634

**Comment by simeetnayan81** [comment]
@aditya02shah This list here is not updated, it's just for reference. So you might have to manually find out which are already migrated.

**Comment by sitamgithub-MSIT** [comment]
@divyashreepathihalli I would like to take up the Efficient Object Detection with YOLOV8 and KerasCV example.

**Comment by divyashreepathihalli** [comment]
> @divyashreepathihalli I would like to take up the Efficient Object Detection with YOLOV8 and KerasCV example.

Thanks @sitamgithub-MSIT - please also take a look at this comment for updating the example - https://github.com/keras-team/keras-cv/issues/2211#issuecomment-1858877591

**Comment by pranavvp16** [comment]
I have migrated the `x_ray_classification_tpu`  example to keras3. [keras-team/keras-io#1756](https://github.com/keras-team/keras-io/pull/1756)

**Comment by sitamgithub-MSIT** [comment]
@divyashreepathihalli Apart from these examples, there are some examples I tried out for which a problem appeared: a missing dataset. For example, the public dataset link does not exist right now. Or if data is taken from the GCP bucket, it does not exist. Should I mark them then anywhere?

**Comment by sitamgithub-MSIT** [comment]
Also, some of our examples, I worked I see the PR already exists, but somehow it's been a month since the contributor added those changes to the example.py file. It is not merged due to the fact that .ipynb and .md files are not generated. What should we do in those cases?

**Comment by divyashreepathihalli** [comment]
> @divyashreepathihalli Apart from these examples, there are some examples I tried out for which a problem appeared: a missing dataset. For example, the public dataset link does not exist right now. Or if data is taken from the GCP bucket, it does not exist. Should I mark them then anywhere?

Yes, I do know about this one, can you share what dataset you are talking about? We might be able to find a mirror site.

**Comment by divyashreepathihalli** [comment]
> Also, some of our examples, I worked I see the PR already exists, but somehow it's been a month since the contributor added those changes to the example.py file. It is not merged due to the fact that .ipynb and .md files are not generated. What should we do in those cases?

Please do respond on that PR that you would like to take it up.

**Comment by sitamgithub-MSIT** [comment]
> > @divyashreepathihalli Apart from these examples, there are some examples I tried out for which a problem appeared: a missing dataset. For example, the public dataset link does not exist right now. Or if data is taken from the GCP bucket, it does not exist. Should I mark them then anywhere?
> 
> Yes, I do know about this one, can you share what dataset you are talking about? We might be able to find a mirror site.

[Classification with Gated Residual and Variable Selection Networks](https://keras.io/examples/structured_data/classification_with_grn_and_vsn/) In this example, the dataset currently does not exist.

[How to train a Keras model on TFRecord files](https://keras.io/examples/keras_recipes/tfrecord/) In this example, the GCS bucket does not exist at present.

**Comment by sitamgithub-MSIT** [comment]
@divyashreepathihalli For earlier examples that were migrated to Keras 3, in some of those examples, some TF ops or numpy ops that were used can actually be substituted with corresponding Keras ops. Should those need any changes then? Because ultimately, the examples are running fine with those unchanged tf or np ops with the backends specified.

**Comment by divyashreepathihalli** [comment]
@sitamgithub-MSIT we should be replaceing all tf ops with keras ops and the functionality should still be the same. I did not quite understand your second question? do you have an example?

**Comment by divyashreepathihalli** [comment]
> [Classification with Gated Residual and Variable Selection Networks](https://keras.io/examples/structured_data/classification_with_grn_and_vsn/) In this example, the dataset currently does not exist.
>found this one  https://archive.ics.uci.edu/static/public/20/census+income.zip
> [How to train a Keras model on TFRecord files](https://keras.io/examples/keras_recipes/tfrecord/) In this example, the GCS bucket does not exist at present. - not sure

**Comment by sitamgithub-MSIT** [comment]
> @sitamgithub-MSIT we should be replaceing all tf ops with keras ops and the functionality should still be the same. I did not quite understand your second question? do you have an example?

For example, this section of code from this keras 3 code example [Simple MNIST convnet](https://keras.io/examples/vision/mnist_convnet/)

```
# Scale images to the [0, 1] range
x_train = x_train.astype("float32") / 255
x_test = x_test.astype("float32") / 255
# Make sure images have shape (28, 28, 1)
x_train = np.expand_dims(x_train, -1)
x_test = np.expand_dims(x_test, -1)
print("x_train shape:", x_train.shape)
print(x_train.shape[0], "train samples")
print(x_test.shape[0], "test samples")
```
Here, ```np.expand_dims``` can be replaced with ```ops.expand_dims```. I tried and worked fine with every backend.

Another one from this keras 3 code example is [Image classification via fine-tuning with EfficientNet](https://keras.io/examples/vision/image_classification_efficientnet_fine_tuning/)

```
size = (IMG_SIZE, IMG_SIZE)
ds_train = ds_train.map(lambda image, label: (tf.image.resize(image, size), label))
ds_test = ds_test.map(lambda image, label: (tf.image.resize(image, size), label))
```
Here is also ```keras.ops.image.resize```, which can be used. 

So, that I am asking, these examples are already marked as keras 3 on the website, so should I change this ops and let it be as it is?

**Comment by sitamgithub-MSIT** [comment]
@divyashreepathihalli for this code ```(x_train, y_train), (x_test, y_test) = keras.datasets.cifar100.load_data()``` getting this error ```Exception: URL fetch failure on https://www.cs.toronto.edu/~kriz/cifar-100-python.tar.gz: 503 -- Service Unavailable```.

**Comment by sitamgithub-MSIT** [comment]
[Data-Parallel Training with KerasNLP and tf.distribute](https://keras.io/examples/nlp/data_parallel_training_with_keras_nlp/) This is an example using a dataset that shows 403: Forbidden. Giving the message "Access Denied.".

Executed Notebook: https://colab.research.google.com/drive/1RbY321lKSodVP3Iy2vVq4kRlTUpfflMp#scrollTo=BfVweZvmOVpI 

cc: @divyashreepathihalli

**Comment by divyashreepathihalli** [comment]
@sitamgithub-MSIT kerasNLP would be a better place for this question.

**Comment by sitamgithub-MSIT** [comment]
@sitamgithub-MSIT KerasNLP would be a better place for this question.

Ok, I will raise an issue then in KerasNLP. Also here in this example [CycleGAN](https://keras.io/examples/generative/cyclegan/), instance normalization is applied, and Keras does not have any support as of now. So, I replaced this with a batch normalization layer in Keras. Another thing is that the author proposed to run the model for only one epoch, as it takes nearly 7 to 8 minutes for one epoch. So I did that and made a few changes to the checkpoint file name so the weights could be stored. But the author here did the inference step with the same model architecture that ran over 90 epochs. As the model is not in.weights.h5 or.keras format, I cannot perform the inferencing. So I need your guidance on this. Also, here is the notebook I ran for reference.

Executed Notebook: https://colab.research.google.com/drive/1NkXKivVipH2XrbbVlDJh0rgUogtr6yKZ?usp=sharing
